pygsti/forwardsims/torchfwdsim.py

```diff
@@ -19,206 +25,240 @@
     from pygsti.layouts.copalayout import CircuitOutcomeProbabilityArrayLayout
     import torch
 
-import numpy as np
 import warnings as warnings
-from collections import OrderedDict
 from pygsti.modelmembers.torchable import Torchable
 from pygsti.forwardsims.forwardsim import ForwardSimulator
 
+try:
+    import torch
+    TORCH_ENABLED = True
+except ImportError:
+    TORCH_ENABLED = False
+    pass
+
 
 class StatelessCircuit:
     """
-    Helper data structure useful for simulating a specific circuit quantum (including prep,
+    Helper data structure useful for simulating a specific quantum circuit (including prep,
     applying a sequence of gates, and applying a POVM to the output of the last gate).
     
     The forward simulation can only be done when we have access to a dict that maps
-    pyGSTi Labels to certain PyTorch Tensors.
+    pyGSTi Labels to certain PyTorch Tensors. Construction of that dict is managed by the
+    StatelessModel class, defined below.
     """
 
     def __init__(self, spc: SeparatePOVMCircuit):
         self.prep_label = spc.circuit_without_povm[0]
-        self.op_labels  = spc.circuit_without_povm[1:]
+        if len(spc.circuit_without_povm) > 1:
+            self.op_labels  = spc.circuit_without_povm[1:]
+        else:
+            # Importing this at the top of the file would create a circular dependency.
+            from pygsti.circuits.circuit import Circuit
+            self.op_labels = Circuit(tuple())
```

Comment:
I have learned a bit more since I originally commented on this line, and it turns out that what you had actually does work. It turns out that, while if you were to `spc.circuit_without_povm[1]` on a length 1 Circuit you'd indeed get an attribute error, when grabbing the slice `spc.circuit_without_povm[1:]` this instead returns the empty circuit. We can debate how sensible that is on a design level (it turns out a lot of the code knows about and uses this fact, however, so changing this behavior would be nontrivial), but it does work here. There is also an argument to be made that this might be slightly preferable in this case, as the sliced circuit inherits the `line_labels` value directly from the parent, which you'd otherwise need to add some logic to grab in creating it directly as above.

So my recommendation would be to revert this back to how you originally had it. Apologies for the misdirection.

Reply:
Man, you've really gotten into the weeds on this stuff, lol. Happy to revert, and no worries!